Un-inline Functions.h into Functions.cpp

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39446 Un-inline Functions.h into Functions.cpp**

In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21862037](https://our.internmc.facebook.com/intern/diff/D21862037)

ezyang committed:
Un-inline Functions.h into Functions.cpp

In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Un-inline Functions.h into Functions.cpp"


In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Un-inline Functions.h into Functions.cpp"


In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

malfet commented on aten/src/ATen/templates/Functions.cpp:
```diff
@@ -0,0 +1,18 @@
+#pragma once
```
One shouldn't have `#pragam once` in .cpp file
```suggestion
```

ezyang committed:
Update on "Un-inline Functions.h into Functions.cpp"


In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ShawnZhong commented:
Just curious, why would un-inlining functions reduce the startup time? Cuz I always thought that inlining is an optimization technique.

ezyang commented:
Startup time of the "compiler" (e.g., how long the compiler takes parsing files before we actually start compiling)

ezyang committed:
Update on "Un-inline Functions.h into Functions.cpp"


In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21862037](https://our.internmc.facebook.com/intern/diff/D21862037)

[ghstack-poisoned]

ezyang committed:
Update on "Un-inline Functions.h into Functions.cpp"


In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21862037](https://our.internmc.facebook.com/intern/diff/D21862037)

[ghstack-poisoned]

ezyang committed:
Update on "Un-inline Functions.h into Functions.cpp"


In my unscientific testing, this reduces startup time by 50% on gcc 8.3.
That's a big fucking deal.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21862037](https://our.internmc.facebook.com/intern/diff/D21862037)

[ghstack-poisoned]
